### PR TITLE
py-tensorflow and in-tree ROCm

### PR DESCRIPTION
**Pull request by aweits**
After a few conversations with @parallelo in slack, this is what is
required to build tensorflow with ROCm in spack. After this work
started, I discovered the existence of
https://github.com/RadeonOpenCompute/rocm-spack. At some point these
two efforts should be combined upstream? Also, I do not currently
have hardware to validate/test this on.
@gkathirv @srekolam @streamhsa @radhaksri @adamjstewart

### CODE REVIEW

File: var/spack/repos/builtin/packages/rocm-hip-clang/package.py [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+
+
+class RocmHipClang(CMakePackage):
+    """HIP-Clang: llvm and clang for ROCm"""
+
+    homepage = "https://github.com/RadeonOpenCompute/llvm-project"
+    url = "https://github.com/RadeonOpenCompute/llvm-project/archive/rocm-3.5.0.tar.gz"
+
+    version('3.5.0', '4878fa85473b24d88edcc89938441edc85d2e8a785e567b7bd7ce274ecc2fd9c')
+    depends_on('gcc')
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/rocm-hip-clang/package.py:14]: I'm not sure how I feel about this. Compilers aren't really modeled as dependencies just yet. Does it only work with GCC?

File: var/spack/repos/builtin/packages/rocm-hip/package.py [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+            filter_file(r'FCC',
+                        'FCC|hipcc',
+                        'hipcc')
+            filter_file(r'command\=\"\$SPACK_CXX\"',
+                        'command="{0}/hipcc.orig"'.format(self.prefix.bin),
+                        'hipcc')
+            filter_file(r'die \"Spack compiler must be run.*',
+                        'command="%s/hipcc.orig"\n'
+                        '\texec "${command}" "$@"' % self.prefix.bin,
+                        'hipcc')
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/rocm-hip/package.py:83]: This also looks sketchy. I've never seen a package copy Spack's compiler wrappers into the installation prefix. Can we filter the compiler wrappers out post-installation?

**Review by adamjstewart — COMMENTED** [review]
I think the TF package is missing a `+rocm` variant.

A few comments about how compilers are handled in the packages.

File: var/spack/repos/builtin/packages/rocm-hip-clang/package.py (hunk 1 above)
Comment (aweits) [reply var/spack/repos/builtin/packages/rocm-hip-clang/package.py:14]: This is my lame attempt to deal with issues surrounding lib(std)c++ and the fact that clang/llvm happily grabs onto the system GCC if not appropriately guided. (I have been looking for something more elegant in this space as well)

**Comment by srekolam** [comment]
The validation for https://github.com/RadeonOpenCompute/rocm-spack is still a work in progress but they are verified to be able to spack  install successfully on a base centos docker. A couple of third party dependencies are seen around the following for which I did not see the spack built-in recipes at this momemt ( gcc ;gcc-c++ ;lbzip2 ;file ; make;numactl-devel ;libbabeltrace-devel) etc.  The spack recipes for some of the rocm math-libs(rocfft,rocblas etc etc used in py-tensorflow) are currently under development .

File: var/spack/repos/builtin/packages/rocm-hip/package.py (hunk 2 above)
Comment (aweits) [reply var/spack/repos/builtin/packages/rocm-hip/package.py:83]: Yep, unfortunately when compiling tensorflow/etc, we need spack's compiler wrappers to do the correct injection of headers when this different compiler is called outside of the wrappers. I'm open to suggestions.

File: var/spack/repos/builtin/packages/rocm-hip-clang/package.py (hunk 1 above)
Comment (aweits) [reply var/spack/repos/builtin/packages/rocm-hip-clang/package.py:14]: Put another way, if we mix libc++ and libstdc++ here we have issues. The "solution"/intent here was to force everything in the chain to be libstdc++.

File: var/spack/repos/builtin/packages/rocm-hip/package.py (hunk 2 above)
Comment (aweits) [reply var/spack/repos/builtin/packages/rocm-hip/package.py:83]: This is also why @parallelo had to revert my addition of external protobuf in order for his build to work here:
https://github.com/parallelo/spack/tree/rocm-update-py-tensorflow-workarounds

Comment (srekolam) [reply var/spack/repos/builtin/packages/rocm-hip/package.py:83]: @aweits  , i made some changes to py-torch recipe to work for rocm. It is currently in  
https://github.com/kolamsrinivas/spack/tree/rocm-update-py-torch

Comment (adamjstewart) [reply var/spack/repos/builtin/packages/rocm-hip/package.py:83]: Please submit a PR for that too, it looks like a much simpler change than what is needed for TF.

File: var/spack/repos/builtin/packages/rocm-rocr-runtime/package.py [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+    kernels to available HSA ROCm kernel agents. Reference source code
+    for the core runtime is also available."""
+
+    homepage = "https://github.com/RadeonOpenCompute/ROCR-Runtime"
+    url = "https://github.com/RadeonOpenCompute/ROCR-Runtime/archive/rocm-3.5.0.tar.gz"
+
+    root_cmakelists_dir = "src"
+
+    version('3.5.0', '52c12eec3e3404c0749c70f156229786ee0c3e6d3c979aed9bbaea500fa1f3b8')
+    depends_on('rocm-roct-thunk-interface')
```
Comment (parallelo) [inline var/spack/repos/builtin/packages/rocm-rocr-runtime/package.py:19]: @aweits Likely also needs a `libelf` dependency here.  

Seeing these types of errors:
```
==> Error: Failed to install rocm-rocr-runtime due to ChildError: ProcessError: Command exited with status 1:
    'cmake' '-G' 'Unix Makefiles' '-DCMAKE_INSTALL_PREFIX:STRING=/root/spack/opt/spack/linux-ubuntu18.04-nehalem/gcc-7.4.0/rocm-rocr-runtime-3.5.0-sas4qfsi3d4qs5uiafhwx44muxiq5jd6' '-DCMAKE_BUILD_TYPE:STRING=RelWithDebInfo' '-DCMAKE_VERBOSE_MAKEFILE:BOOL=ON' '-DCMAKE_INS
TALL_RPATH_USE_LINK_PATH:BOOL=OFF' '-DCMAKE_INSTALL_RPATH:STRING=/root/spack/opt/spack/linux-ubuntu18.04-nehalem/gcc-7.4.0/rocm-rocr-runtime-3.5.0-sas4qfsi3d4qs5uiafhwx44muxiq5jd6/lib;/root/spack/opt/spack/linux-ubuntu18.04-nehalem/gcc-7.4.0/rocm-rocr-runtime-3.5.0-sas4q
fsi3d4qs5uiafhwx44muxiq5jd6/lib64;/root/spack/opt/spack/linux-ubuntu18.04-nehalem/gcc-7.4.0/rocm-roct-thunk-interface-3.5.0-unyl6gjbinjcdmwoz4jqhv7472yevfiq/lib' '-DCMAKE_PREFIX_PATH:STRING=/root/spack/opt/spack/linux-ubuntu18.04-nehalem/gcc-7.4.0/cmake-3.17.1-2dczyfsidp
vwsmcr5cnoqhepagk4lqfu;/root/spack/opt/spack/linux-ubuntu18.04-nehalem/gcc-7.4.0/rocm-roct-thunk-interface-3.5.0-unyl6gjbinjcdmwoz4jqhv7472yevfiq' '-DCMAKE_BUILD_WITH_INSTALL_RPATH=1' '/tmp/root/spack-stage/spack-stage-rocm-rocr-runtime-3.5.0-sas4qfsi3d4qs5uiafhwx44muxiq
5jd6/spack-src/src'
1 error found in build log:
     11    -- Check for working CXX compiler: /root/spack/lib/spack/env/gcc/g++
     12    -- Check for working CXX compiler: /root/spack/lib/spack/env/gcc/g++
            - works
     13    -- Detecting CXX compiler ABI info
     14    -- Detecting CXX compiler ABI info - done
     15    -- Detecting CXX compile features
     16    -- Detecting CXX compile features - done
  >> 17    CMake Error at /root/spack/opt/spack/linux-ubuntu18.04-nehalem/gcc-7
           .4.0/cmake-3.17.1-2dczyfsidpvwsmcr5cnoqhepagk4lqfu/share/cmake-3.17/
           Modules/FindPackageHandleStandardArgs.cmake:164 (message):
     18      Could NOT find LibElf (missing: LIBELF_LIBRARIES LIBELF_INCLUDE_DI
           RS)
     19    Call Stack (most recent call first):
     20      /root/spack/opt/spack/linux-ubuntu18.04-nehalem/gcc-7.4.0/cmake-3.
           17.1-2dczyfsidpvwsmcr5cnoqhepagk4lqfu/share/cmake-3.17/Modules/FindP
           ackageHandleStandardArgs.cmake:445 (_FPHSA_FAILURE_MESSAGE)
     21      cmake_modules/FindLibElf.cmake:49 (FIND_PACKAGE_HANDLE_STANDARD_AR
           GS)
     22      CMakeLists.txt:72 (find_package)
```

Comment (kolamsrinivas) [reply var/spack/repos/builtin/packages/rocm-rocr-runtime/package.py:19]: @aweits ,  you can use the packages that are already built and tested with rocm-3.5.1 . it is available here at https://github.com/RadeonOpenCompute/rocm-spack/tree/develop/packages . We are looking to raise a PR for these soon

Comment (aweits) [reply var/spack/repos/builtin/packages/rocm-rocr-runtime/package.py:19]: I'll be removing the ROCm components from the PR, and we'll go with the packages that the ROCm team has opened PR's for.

**Comment by adamjstewart** [comment]
@aweits what's the status of this PR now that some of the ROCm dependencies have been merged? Is anything else needed on the ROCm side?

**Comment by aweits** [comment]
> @aweits what's the status of this PR now that some of the ROCm dependencies have been merged? Is anything else needed on the ROCm side?

I'll look at this later tonight and see what's required...

**Comment by haampie** [comment]
Should this be closed?

**Comment by aweits** [comment]
> Should this be closed?

I can pick this up again, has the ROCm stuff finally settled down a bit?

**Comment by haampie** [comment]
Yes, it's quite stable :)

**Comment by alalazo** [comment]
Closing as stale, but feel free to reopen if you want to continue working on the PR.

**Pull request closed without merging**